### PR TITLE
docs: add tll tutorial

### PR DESCRIPTION
**Pull request by bwanglzu**

### CODE REVIEW

File: docs/get-started/totally-looks-like/index.md [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+2. we consider the left image in each pair as `anchor`, right image in each pair as `positive`.
+3. for the positive image, we assign the tag `finetuner_label` as 1 to tell Finetuner it's a positive instance.
+4. we randomly sample 1 document in `right_da` as a negative image, and we assign `finetuner_label` as `-1` indicates it is a negative instance.
+5. we attach `positive` and `negative` instance as `matches` of the `anchor`.
+
+It should be noted that:
+
+1. we must make sure our randomly sample negative image is not itself.
+2. in practice, you could sample multiple negative images, such as 1 positive 4 negatives.
+3. in practice, you should select hard negatives, for more information, please refer to [losses and miners](https://finetuner.jina.ai/components/tuner/loss/).
```
Comment (tadejsv) [inline docs/get-started/totally-looks-like/index.md:70]: In fact, why not use the miners directly? Why mention them here as "in practice" - they are not hard to use, it's just one line of code. Have you checked if they work well? If they do, all this thing with manually building matches can be skipped, which would be great

File: docs/get-started/totally-looks-like/index.md [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+```
+
+In this tutorial, we rely on `torchvision` to get a pre-trained ResNet50 on ImageNet dataset.
+We'll need to install jina to prepare training/test data since it is the expected input of Finetuner.
+Last but not least, we import finetuner and the loss function `TripletLoss`.
+
+Afterwards, we load all images from unzipped `left` and `right` folder and turn them into sorted order.
+
+```python
+from jina import DocumentArray
```
Comment (JoanFM) [inline docs/get-started/totally-looks-like/index.md:39]: docarray

File: docs/get-started/totally-looks-like/index.md (hunk 1 above)
Comment (JoanFM) [reply docs/get-started/totally-looks-like/index.md:70]: I think is good to show this process for learning in any case.

Comment (bwanglzu) [reply docs/get-started/totally-looks-like/index.md]: sure, will update

File: docs/get-started/totally-looks-like/index.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
+# Finetuning ResNet50 on Totally Looks Like Dataset
+
+In this tutorial, we will use Finetuner to finetune embeddings on [Totally Looks Like dataset](https://sites.google.com/view/totally-looks-like-dataset) (TLL).
+
+>Totally-Looks-Like is a dataset and benchmark challenging machine-learned representations to reproduce human perception of image similarity. 
+
+The dataset consist 6016 pairs of images (12032 in total).
```
Comment (nan-wang) [inline docs/get-started/totally-looks-like/index.md:7]:
```suggestion
The dataset consists 6016 pairs of images (12032 in total).
```

File: docs/get-started/totally-looks-like/index.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+And we use Finetuner to tune the model to get a better embedding. But how it works?
+
+Finetuner adopts the idea of `transfer learning` and `metric learning`. The rationale is
+1. TLL dataset is a relative small dataset, it's not reasonable to train a large network, such as ResNet on TLL from scratch. So we freeze the weights of a pre-trained ResNet and train a MLP head to adopt our new dataset.
+2. TLL dataset consist pairs of images which can be formed as a positive pair, and a random image can can be considered as a negative pair. We can form a `triplet` and use the Finetuner `TripletLoss`. We expect after fine-tuning, the distance between positive pairs can be pull closer, while the distance between positive and negative pairs can be push away.
```
Comment (nan-wang) [inline docs/get-started/totally-looks-like/index.md:12]:
```suggestion
2. TLL dataset consists pairs of images which can be formed as a positive pair, and a random image can can be considered as a negative pair. We can form a `triplet` and use the `TripletLoss`. We expect after fine-tuning, the distance between positive pairs get closer than negative pairs.
```

**Review by nan-wang — CHANGES_REQUESTED** [review]

**Comment by bwanglzu** [comment]
still re-writing the content, will let you know when it's ready to review

File: docs/get-started/totally-looks-like/index.md [hunk 5]
```diff
@@ -0,0 +1,206 @@
...
+## Environment & Data Preparation
+
+We will download `left.zip` and `right.zip`, as stated before,
+each of them consists of 6016 images which can be formed into pairs based on the same file name.
+
+
+```shell
+pip install gdown
+pip install finetuner
+pip install torchvision
```
Comment (nan-wang) [inline docs/get-started/totally-looks-like/index.md:21]:
```suggestion
pip install torchvision
pip install torch>=1.9.0
```

Comment (bwanglzu) [reply docs/get-started/totally-looks-like/index.md:24]: i think the latest torchvision will take care the torch version so I did not add it

**Comment by tadejsv** [comment]
What performance do you get without freezing any layers?

**Comment by bwanglzu** [comment]
@tadejsv I havn't try given so many combinations of parameters, you can make a copy and give a try.

File: docs/get-started/totally-looks-like/index.md (hunk 5 above)
Comment (nan-wang) [reply docs/get-started/totally-looks-like/index.md:25]: I see. The reason is due to my `pip` is not up-to-date

**Review by nan-wang — COMMENTED** [review]
It would be better if we can add a comparison of `hit@k` before and after finetuning

File: docs/get-started/totally-looks-like/index.md [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+right_da = DocumentArray.from_files('right/*.jpg')
+# we use 80% for training machine learning model.
+left_da.sort(key=lambda x: x.uri)
+right_da.sort(key=lambda x: x.uri)
+
+ratio = 0.8
+train_size = int(ratio * len(left_da))
+
+train_da = left_da[:train_size] + right_da[:train_size]
+train_da = train_da.shuffle()
```
Comment (makram93) [inline docs/get-started/totally-looks-like/index.md:51]: Shuffling just the train data does not make sense. Can we shuffle the entire set and take a subset of it as training data?

```suggestion

ratio = 0.8
train_size = int(ratio * len(left_da))

# shuffling the left_da and right_da in unison
idx = np.arange(len(left_da))
np.random.shuffle(idx)
left_da = left_da[list(idx)]
right_da = right_da[list(idx)]

train_da = left_da[:train_size] + right_da[:train_size]
```

**Comment by bwanglzu** [comment]
@nan-wang it's added, refresh the page.

File: docs/get-started/totally-looks-like/index.md [hunk 7]
```diff
@@ -0,0 +1,198 @@
+# Finetuning ResNet50 on Totally Looks Like Dataset
+
+In this tutorial, we will use Finetuner to finetune embeddings on [Totally Looks Like dataset](https://sites.google.com/view/totally-looks-like-dataset) (TLL) dataset.
+The dataset consists of 6016 pairs of images (12032 in total).
+
+>Totally-Looks-Like is a dataset and benchmark challenging machine-learned representations to reproduce human perception of image similarity.
+
+![overview](overview.png)
```
Comment (nan-wang) [inline docs/get-started/totally-looks-like/index.md:8]:
````suggestion
```{figure} overview.png
:width: 70%
```
````

File: docs/get-started/totally-looks-like/index.md [hunk 8]
```diff
@@ -0,0 +1,198 @@
+# Finetuning ResNet50 on Totally Looks Like Dataset
+
+In this tutorial, we will use Finetuner to finetune embeddings on [Totally Looks Like dataset](https://sites.google.com/view/totally-looks-like-dataset) (TLL) dataset.
+The dataset consists of 6016 pairs of images (12032 in total).
+
+>Totally-Looks-Like is a dataset and benchmark challenging machine-learned representations to reproduce human perception of image similarity.
```
Comment (nan-wang) [inline docs/get-started/totally-looks-like/index.md:6]:
````suggestion
```{tip}
Totally-Looks-Like is a dataset and benchmark challenging machine-learned representations to reproduce human perception of image similarity. As shown below, each image patch in the left has a corresponded similar image patch in the right. 
```
````

File: docs/get-started/totally-looks-like/index.md [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+
+![overview](overview.png)
+
+*Each image patch in the left has a corresponded similar image patch in the right.*
```
Comment (nan-wang) [inline docs/get-started/totally-looks-like/index.md:10]:
```suggestion
```

File: docs/get-started/totally-looks-like/index.md [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+
+Finetuner adopts the idea of [transfer learning](https://en.wikipedia.org/wiki/Transfer_learning) and [metric learning](https://en.wikipedia.org/wiki/Similarity_learning#Metric_learning). The rationale is
```
Comment (nan-wang) [inline docs/get-started/totally-looks-like/index.md:12]:
```suggestion
The rationale for choosing TLL dataset is
```

File: docs/get-started/totally-looks-like/index.md [hunk 11]
```diff
@@ -0,0 +1,198 @@
...
+1. TLL dataset is a relatively small dataset, it's not reasonable to train a large network, such as ResNet on TLL from scratch. So we train part of the network by freezing layers.
+2. TLL dataset consists of pairs of images that can be formed as a positive pair (same classes), and a random image can be considered as a negative pair. We can form a `triplet` and use the Finetuner `TripletLoss`. We expect after fine-tuning, the distance between positive pairs can be pulled closer, while the distance between positive and negative pairs can be pushed away.
```
Comment (nan-wang) [inline docs/get-started/totally-looks-like/index.md:14]:
```suggestion
- it is a relatively small dataset and doesn't require training a large network from scratch. So we train only part of the ResNet-50 by freezing layers.
- it consists of pairs of images that can be formed as positive pairs (same classes). A random image can be considered as a negative pair. We construct `triplet` and use the `TripletLoss`. 

After fine-tuning, the distance between positive pairs is expected to be pulled closer, while the distance between positive and negative pairs is expected to be pushed away.
```

File: docs/get-started/totally-looks-like/index.md (hunk 6 above)
Comment (nan-wang) [reply docs/get-started/totally-looks-like/index.md:51]: i think it is fine not a big deal

**Review by nan-wang — APPROVED** [review]
LGTM👍

File: docs/get-started/totally-looks-like/index.md [hunk 12]
```diff
@@ -0,0 +1,203 @@
...
+right_da = DocumentArray.from_files('right/*.jpg')
+# we use 80% for training machine learning model.
+left_da.sort(key=lambda x: x.uri)
+right_da.sort(key=lambda x: x.uri)
+
+ratio = 0.8
+train_size = int(ratio * len(left_da))
+
+train_da = left_da[:train_size] + right_da[:train_size]
+train_da = train_da.shuffle()
```
Comment (tadejsv) [inline docs/get-started/totally-looks-like/index.md:56]: I would remove this shuffle as it server no purpose. Real shuffling will be done by the class dataset

File: docs/get-started/totally-looks-like/index.md (hunk 6 above)
Comment (bwanglzu) [reply docs/get-started/totally-looks-like/index.md]: @makram93 I'll remove shuffle since finetuner will do the job
